### ISSUE TITLE
Interactive transaction is not rolling back if it's interrupted in the middle by "throw Error"

### ISSUE THREAD

**Issue by oyatek** [prisma/prisma#11107]
### Bug description

I have an interactive transaction that is almost identical to the official example. Decrement the counter, check if it's below zero and throw an error, if it's > 0 then create an object. The thing is that even though I do `throw new Error()` when the counter is < 0 (and I receive the error message in the browser), the transaction is not getting rolled back - and I end up with negative counters in the database, which should not happen. Why is it not rolling back?

The code:

```
  const order = await db.$transaction(async (prisma) => {

    // 1. Decrement amount of tickets in the Game
    const _game = await db.game.update({ where: {id: game.id}, data: {
      ticketsNum: {
        decrement: input.tickets.length,
      }}
    })

    // 2. Verify that the amount of tickets didn't go below zero.
    if (_game.ticketsNum < 0) {
      throw new Error(`The requested number of tickets is not available`)
    }

    // 3. Create new order with tickets and return it
    return db.order.create({ data: data })
  })
```

If it matters, I'm using Prisma 2 as part of Blitz.js installation

### How to reproduce

Create an interactive transaction and add `throw Error()` in the middle.

### Expected behavior

Transaction must rollback when an error is thrown in the middle of it

### Prisma information

---

### Environment & setup

- OS: Mac OS 
- Database: PostgreSQL
- Node.js version: 17

### Prisma Version

```
prisma                  : 3.7.0
@prisma/client          : 3.7.0
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine 8746e055198f517658c08a0c426c7eec87f5a85f (at node_modules/prisma/node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Migration Engine        : migration-engine-cli 8746e055198f517658c08a0c426c7eec87f5a85f (at node_modules/prisma/node_modules/@prisma/engines/migration-engine-darwin-arm64)
Introspection Engine    : introspection-core 8746e055198f517658c08a0c426c7eec87f5a85f (at node_modules/prisma/node_modules/@prisma/engines/introspection-engine-darwin-arm64)
Format Binary           : prisma-fmt 8746e055198f517658c08a0c426c7eec87f5a85f (at node_modules/prisma/node_modules/@prisma/engines/prisma-fmt-darwin-arm64)
Default Engines Hash    : 8746e055198f517658c08a0c426c7eec87f5a85f
Studio                  : 0.445.0
Preview Features        : interactiveTransactions

```

**Comment by Bearfinn** [comment]
I am not sure if this is right, but I believe you have to use `prisma.xxx` inside $transaction callback.
The main `db` might not contain information and did not initiate transaction

**Comment by oyatek** [comment]
@Bearfinn it might be the case, thank for pointing this out! I will try to change the code and get back with the results.

**Comment by matthewmueller** [comment]
Could you try updating to 3.9.0 and trying again? We fixed a lot of bugs in 3.10.0 and it seems like this might have been resolved.

**Comment by millsp** [comment]
We have fixed some rollback issues in `3.9.0`, but it your case the problem lies with `await db.$transaction(async (prisma) => {`. When you open a transaction, you *must* use the provided `prisma` client via the callback. This parameter is your prisma client but annotated with a transaction id. If you don't, you are performing a non-transactional query in the middle of your interactive transaction. I'm closing this, but feel free to ping me if needed. Thanks!

**Comment by oyatek** [comment]
Thank you very much for the explanation
